turbine/src/addr_cache.rs

```diff
@@ -0,0 +1,412 @@
+    }
+
+    // Stores (root-distance, socket-addresses) precomputed speculatively for
+    // the given shred-id.
+    pub(crate) fn put(
+        &mut self,
+        shred: &ShredId,
+        entry: (/*root_distance:*/ u8, Box<[SocketAddr]>),
+    ) {
+        self.get_cache_entry_mut(shred.slot())
+            .put(shred.shred_type(), shred.index(), entry);
+        self.maybe_trim_cache();
+    }
+
+    // Records data observed from incoming shreds at retransmit stage.
+    pub(crate) fn record(&mut self, slot: Slot, stats: &mut RetransmitSlotStats) {
+        // All addresses should be for the same slot.
+        debug_assert!(stats.addrs.iter().all(|(shred, _, _)| shred.slot() == slot));
+        // Update rolling window count of shreds per slot.
+        let num_shreds: usize = stats.num_shreds_received.iter().sum();
+        if num_shreds > 0 {
+            self.num_shreds += num_shreds;
+            self.window.push_back((slot, num_shreds));
+            *self.counts.entry(slot).or_default() += num_shreds;
+            self.maybe_trim_slot_counts();
+        }
+        debug_assert!(self.verify());
+        // If there are no addresses to cache and the cache entry is not
+        // allocated for the slot yet, then ignore.
+        if stats.addrs.is_empty() && !self.cache.contains_key(&slot) {
+            return;
+        }
+        // Update the cached entry for the slot.
+        let entry = self.get_cache_entry_mut(slot);
+        entry.max_index_code = entry.max_index_code.max(stats.max_index_code);
+        entry.max_index_data = entry.max_index_data.max(stats.max_index_data);
+        entry.last_shred_in_slot |= stats.last_shred_in_slot;
+        for (shred, root_distance, addrs) in std::mem::take(&mut stats.addrs) {
+            debug_assert_eq!(shred.slot(), slot);
+            entry.put(shred.shred_type(), shred.index(), (root_distance, addrs));
+        }
+        self.maybe_trim_cache();
+        debug_assert!(self.verify());
+    }
+
+    // Returns num_shreds shred-ids to speculatively pre-compute turbine tree.
+    // ShredIds are chosen based on which slots have received most number of
+    // shreds within the rolling window.
+    pub(crate) fn get_shreds(&mut self, num_shreds: usize) -> Vec<ShredId> {
+        fn make_shred(slot: Slot, (shred_type, index): (ShredType, usize)) -> ShredId {
+            ShredId::new(slot, index as u32, shred_type)
+        }
+        if self.counts.len() == 1 {
+            let slot = self.counts.keys().next().copied().unwrap();
+            return self
+                .get_cache_entry_mut(slot)
+                .get_shreds(EXTEND_BUFFER)
+                .take(num_shreds)
+                .map(|entry| make_shred(slot, entry))
+                .collect();
+        }
+        let mut counts: Vec<(/*count:*/ usize, Slot)> = self
+            .counts
+            .iter()
+            .map(|(&slot, &count)| (count, slot))
+            .collect();
+        counts.sort_unstable();
```

Comment:
Does it make sense to bias by how big the cache for this slot is as well? For instance if we initially see `1000` shreds for the current, and there is some network delay we might end up computing many thousands of shreds for for the slot

Reply:
Also thoughts on precomputing future slots? Based on the average shred count per block over our window it might be more beneficial to stop precomputing for a slot once its reached a certain size and instead speculate on upcoming slots.

Reply:
> we might end up wastefully computing many thousands of shreds for for the slot

The cache only precomputes up to `EXTEND_BUFFER` beyond the observed `max_index_{code,data}`.

> Also thoughts on precomputing future slots?

Yeah, we can try out those ideas and see what works.
For now I just started with some basic heuristics.

I thing having some basic core functionality in-place would make it easier to iterate and test other ideas.